src/schema/date_time_options.rs

```diff
@@ -0,0 +1,606 @@
+{
+    fn from(head_tail: SchemaFlagList<Head, Tail>) -> Self {
+        Self::from(head_tail.head) | Self::from(head_tail.tail)
+    }
+}
+
+// Parser function types
+type StringDateTimeParser = Arc<dyn Fn(&str) -> Result<OffsetDateTime, String> + Sync + Send>;
+type NumberDateTimeParser = Arc<dyn Fn(i64) -> Result<OffsetDateTime, String> + Sync + Send>;
+
+/// A struct for holding datetime parsing functions.
+#[derive(Clone)]
+pub struct DateTimeParsersHolder {
+    /// Functions for parsing string input as datetime.
+    string_parsers: VecDeque<StringDateTimeParser>,
+    /// Function for parsing number input as timestamp.
+    number_parser: NumberDateTimeParser,
+}
+
+impl Default for DateTimeParsersHolder {
+    fn default() -> Self {
+        Self {
+            string_parsers: VecDeque::new(),
+            number_parser: make_timestamp_parser(DateTimePrecision::Milliseconds),
+        }
+    }
+}
+
+impl DateTimeParsersHolder {
+    /// Parses string input.
+    pub fn parse_string(&mut self, value: String) -> Result<OffsetDateTime, String> {
+        for (index, parser) in self.string_parsers.iter().enumerate() {
+            if let Ok(date_time) = parser(&value) {
+                // Move successful parser in front of queue.
+                self.string_parsers.swap(0, index);
+                return Ok(date_time);
+            }
+        }
+        Err("Could not parse a valid datetime using all specified parsers.".to_string())
+    }
+
+    /// Parses number input.
+    pub fn parse_number(&self, value: i64) -> Result<OffsetDateTime, String> {
+        (self.number_parser)(value)
+    }
+}
+
+impl From<DateTimeOptions> for DateTimeParsersHolder {
+    fn from(opts: DateTimeOptions) -> Self {
+        let mut string_parsers: VecDeque<StringDateTimeParser> = VecDeque::new();
+        let mut number_parser: NumberDateTimeParser =
+            make_timestamp_parser(DateTimePrecision::Milliseconds);
+        for input_format in opts.input_formats {
+            match input_format {
+                DateTimeFormat::RCF3339 => string_parsers.push_back(Arc::new(rfc3339_parser)),
+                DateTimeFormat::RFC2822 => string_parsers.push_back(Arc::new(rfc2822_parser)),
+                DateTimeFormat::ISO8601 => string_parsers.push_back(Arc::new(iso8601_parser)),
+                DateTimeFormat::Strftime(str_format) => {
+                    string_parsers.push_back(make_strftime_parser(str_format))
+                }
+                DateTimeFormat::Timestamp(precision) => {
+                    number_parser = make_timestamp_parser(precision)
+                }
+            }
+        }
+        DateTimeParsersHolder {
+            string_parsers,
+            number_parser,
+        }
+    }
+}
+
+/// Parses datetime strings using RFC3339 formatting.
+fn rfc3339_parser(value: &str) -> Result<OffsetDateTime, String> {
+    OffsetDateTime::parse(value, &Rfc3339).map_err(|error| error.to_string())
```

Comment:
nitpick: I think you would be better off keepign the original error in these helper functions and convert them on the caller side.